### PR TITLE
Implement new signing spec

### PR DESCRIPTION
**Pull request by qmuntal**
This PR updates the `notation-azure-kv` implementation to conform to the new plugin extension spec using the `GENERATE_SIGNATURE` workflow.

It also improves the authorization mechanism, allowing to define custom environments, not only `azure.PublicCloud`, and to authorize using the following methods:

1. Client credentials: AZURE_TENANT_ID, AZURE_CLIENT_ID and AZURE_CLIENT_SECRET envs
2. Client certificate: AZURE_TENANT_ID, AZURE_CLIENT_ID, AZURE_CERTIFICATE_PATH and AZURE_CERTIFICATE_PASSWORD envs
3. Username password: AZURE_TENANT_ID, AZURE_CLIENT_ID, AZURE_USERNAME and AZURE_PASSWORD
4. MSI, aka Managed Service Identities
5. Azure CLI 2.0, aka `az login`

Signed-off-by: qmuntal <qmuntaldiaz@microsoft.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,74 +0,0 @@
-# referenced from https://github.com/notaryproject/notation/blob/main/Makefile
```
Comment (dtzar) [inline Makefile:1]: Can you please either add this file back OR add instructions on how to build the binary?  

Great that we have the GitHub testing/release/build/lint/etc, but developers need to know how to locally run what is running in CI.

**Comment by dtzar** [comment]
@qmuntal @SteveLasker hitting https://github.com/Azure/azure-cli/issues/22575 with the certificate download using the latest azure-cli and these steps.  Any ideas?  I've given myself all the permissions in policy and I'm the owner of the AKV.

**Comment by qmuntal** [comment]
> @qmuntal @SteveLasker hitting [Azure/azure-cli#22575](https://github.com/Azure/azure-cli/issues/22575) with the certificate download using the latest azure-cli and these steps. Any ideas? I've given myself all the permissions in policy and I'm the owner of the AKV.

I use this command to download the certificate:

`az keyvault certificate download --file $CERT_PATH --id $KEY_ID`

Notice that  I"m just using the `--id` parameter, which matches the notary key id.

File: Makefile (hunk 1 above)
Comment (qmuntal) [reply Makefile:1]: Done

**Comment by aramase** [comment]
> @qmuntal @SteveLasker hitting https://github.com/Azure/azure-cli/issues/22575 with the certificate download using the latest azure-cli and these steps. Any ideas? I've given myself all the permissions in policy and I'm the owner of the AKV.

Added a comment in the issue: https://github.com/Azure/azure-cli/issues/22575#issuecomment-1136285289. I think @qmuntal posted the same here. The doc needs to be updated to use `--id` instead of `--name` with the id.

**Comment by qmuntal** [comment]
> @qmuntal I don't think we can use the key id for --id. The id needs to be the certificate id.

> az keyvault certificate download --file testfile --id https://kindkv.vault.azure.net/keys/notary/<version>
invalid id: The collection "keys" does not match the expected "certificates"

You are right!

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-25s\033[0m %s\n", $$1, $$2}'
+
+.PHONY: all
+all: build
+
+.PHONY: FORCE
+FORCE:
+
+bin/%: cmd/% FORCE
+	/home/qmuntaldiaz/go/bin/go1.18.1 build $(GO_BUILD_FLAGS) -o $@ ./$<
```
Comment (dtzar) [inline Makefile:24]: Please fix `/home/qmuntaldiaz/go/bin/go1.18.1`

**Comment by dtzar** [comment]
Unfortunately, I'm still getting this error:

```
notation sign $IMAGE --key $KEY_NAME --debug
2022/05/24 15:35:04 signing key not found
```

Using fresh built binaries for notary from main branch and notation-azure-kv plugins from this PR
notation version v0.7.1-alpha.1+c61102cdb414a9aa55980bd113496abef3fcf784.dirty
azure-kv   v0.1.0-alpha.1+07758764ba0319d6cf1f17b189b1933e7639e0f3.dirty

Tried getting vscode debugging setup on **notation**, but having trouble getting it to run right with the switches/plugins.
Perhaps @qmuntal we can setup a code share/debug session to figure out what's broken?

**Comment by qmuntal** [comment]
> Unfortunately, I'm still getting this error: 
Using fresh built binaries for notary from main branch and notation-azure-kv plugins from this PR
notation version v0.7.1-alpha.1+c61102cdb414a9aa55980bd113496abef3fcf784.dirty
azure-kv v0.1.0-alpha.1+07758764ba0319d6cf1f17b189b1933e7639e0f3.dirty

It works for me using those binary versions. My take is that $KEY_NAME is not properly set up. Could you run `notation key list` and check the key is available and associated to this plugin?

> Perhaps @qmuntal we can setup a code share/debug session to figure out what's broken?

Sure.

**Comment by dtzar** [comment]
Yep, both are there. 

```shell
notation key ls
NAME             KEY PATH   CERTIFICATE PATH   ID                                                                                        PLUGIN NAME   
wabbitnetworks                                 https://<redacted>.vault.azure.net/keys/wabbitnetworks/2de43313f57f489581f6e0eec4e7ddc2   azure-kv  
notation certificate ls
NAME             PATH                                                             
wabbitnetworks   /mnt/c/src/github/dtzar/azure-docs-pr-1/log/wabbitnetworks.pem
```

**Review by dtzar — APPROVED** [review]

**Review by SteveLasker — APPROVED** [review]
LGTM as I tested this with the forks @qmuntal is developing.

**Pull request merged**